File: surf-api-bukkit/surf-api-bukkit-api/src/main/java/dev/slne/surf/surfapi/bukkit/api/nms/listener/NmsServerboundPacketListener.java
package dev.slne.surf.surfapi.bukkit.api.nms.listener;

import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution;
import dev.slne.surf.surfapi.bukkit.api.nms.listener.packets.NmsPacket;
import dev.slne.surf.surfapi.bukkit.api.packet.listener.listener.PacketListenerResult;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.ApiStatus.OverrideOnly;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NmsUseWithCaution
@NullMarked
public interface NmsServerboundPacketListener<Packet extends NmsPacket> extends
        NmsPacketListener<Packet> {

    @OverrideOnly
    PacketListenerResult handleServerboundPacket(Packet packet, Player player);

    @OverrideOnly
    default PacketListenerResult handleEarlyServerboundPacket(Packet packet, @Nullable Player player) {
        if (player != null) {
            return handleServerboundPacket(packet, player);
        } else {
            return PacketListenerResult.CONTINUE;
        }
    }
}

File: surf-api-bukkit/surf-api-bukkit-api/src/main/kotlin/dev/slne/surf/surfapi/bukkit/api/nms/listener/NmsServerboundPacketListener.kt
package dev.slne.surf.surfapi.bukkit.api.nms.listener

import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution
import dev.slne.surf.surfapi.bukkit.api.nms.listener.packets.serverbound.NmsServerboundPacket
import dev.slne.surf.surfapi.bukkit.api.packet.listener.listener.PacketListenerResult
import org.bukkit.entity.Player
import org.jetbrains.annotations.ApiStatus

@NmsUseWithCaution
interface NmsServerboundPacketListener<Packet : NmsServerboundPacket> : NmsPacketListener<Packet> {
    @ApiStatus.OverrideOnly
    fun handleServerboundPacket(packet: Packet, player: Player): PacketListenerResult
}